README.md
# Intel 8080 Emulator

A Clojure library that can be used to emulate the behavior of an Intel 8080 microprocessor.
Pass in the memory and IO ports and let it run!

WARNING!! This is work in progress! WARNING!!


## Origins

I have been toying with Clojure and have built several small programs with it. 
As much fun as it was, it did not help me in better understanding the power of functional languages.
So I decided to build something more substantial and thought that a microprocessor emulator would be a good project.
There should be enough repetition in it that would require some structure and I am curious about the challenge
of using stateless structures to represent what is essentially a mutable state machine.

As for the choice of processor to emulate, I wanted something small enough to be implemented in a reasonable amount of time.
I built Intel8080-based systems when I was back in college, so it was a natural candidate.
And I have to add that re-reading the microprocessor manual after such a long time was a blast!






## License

Copyright © 2023 santoro63@yahoo.com

This program and the accompanying materials are made available under the
terms of the GNU Lesser General Public License v3, which is available at
https://www.gnu.org/licenses/lgpl-3.0.html


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-02-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-02-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/emulators/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/emulators/compare/0.1.0...0.1.1


test/emulators/core_test.clj
(ns emulators.core-test
  (:require [clojure.test :refer :all]
            [emulators.core :refer :all]))

(deftest a-test
  (testing "FIXED"
    (is (= 1 1))))


test/emulators/intel8080_test.clj
(ns emulators.intel8080-test
  (:require [emulators.intel8080 :as sut]
            [clojure.test :as t]))


(t/deftest data-movement-tests

  (t/testing "register-to-register"
    (let [ r (assoc sut/init-registers :B 34)
          m [ 0x48 0x00 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io)]
      (t/is (= 34 (r-out :C)))
      (t/is (= 01 (r-out :PC)))
      ))

  (t/testing "memory-to-register"
    (let [ r (assoc sut/init-registers :H 0x00 :L 0x05)
          m [ 0x56 0x00 0x02 0x03 0x04 0x05 ]
          io [ ]
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x05 (r-out :D)))
      (t/is (= 0x01 (r-out :PC)))
      ))

  (t/testing "register to memory"
    (let [ r (assoc sut/init-registers :E 22 :H 0x00 :L 0x05)
          m [0x73 0x00 0x00 0x00 0x00 0x00 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 22 (mem-out 0x05)))
      (t/is (= 0x01 (r-out :PC)))
      ))

  (t/testing "immediate to register"
    (let [ r sut/init-registers
          m [ 0x3E 0xFF 0x01 0x02 ]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 255 (r-out :A)))
      (t/is (= 0x02 (r-out :PC)))
      ))

  (t/testing "immediate to memory"
    (let [ r (assoc sut/init-registers :H 0x00 :L 0x04)
          m [ 0x36 0xFF 0x01 0x02 0x03]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 255 (mem-out 0x04)))
      (t/is (= 0x02 (r-out :PC)))
      ))

  (t/testing "load register pair"
    (let [ r sut/init-registers
          m [ 0x21 0x44 0x21 0x00 ]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x44 (r-out :L)))
      (t/is (= 0x21 (r-out :H)))
      )
    (let [ r sut/init-registers
          m [ 0x31 0x44 0x21 0x00 ]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x2144 (r-out :SP)))
      ))

  (t/testing "load accumulator from memory"
    (let [ r sut/init-registers
          m [ 0x3A 0x05 0x00 0x00 0x00 0x14 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x14 (r-out :A)))
      ))

  (t/testing "load memory from accumulator"
    (let [ r (assoc sut/init-registers :A 0x57)
          m [ 0x32 0x05 0x00 0x00 0x00 0x14 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x57 (mem-out 0x05)))
      ))

  (t/testing "load HL direct"
    (let [ r sut/init-registers
          m [ 0x2A 0x05 0x00 0x00 0x00 0x18 0x71 ]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x18 (r-out :L)))
      (t/is (= 0x71 (r-out :H)))
      ))

  (t/testing "load mem from HL"
    (let [ r (assoc sut/init-registers :L 0x33 :H 0x22)
          m [ 0x22 0x04 0x00 0x00 0x00 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x03 (r-out :PC)))
      (t/is (= 0x33 (mem-out 0x04)))
      (t/is (= 0x22 (mem-out 0x05)))
      ))

  (t/testing "load mem(rp) to acc"
    (let [ r (assoc sut/init-registers :B 0x00 :C 0x04)
          m [0x0A 0x00 0x00 0x00 0xFF 0xDD]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x01 (r-out :PC)))
      (t/is (= 0xFF (r-out :A)))
      ))

  (t/testing "load mem(rp) from acc"
    (let [ r (assoc sut/init-registers :A 0xEF :D 0x00 :E 0x03)
          m [0x12 0x00 0x00 0x00 0x00]
          io []
          [r-out mem-out io-out] (sut/step r m io) ]
      (t/is (= 0x01 (r-out :PC)))
      (t/is (= 0xEF (mem-out 0x0003)))
      ))

  (t/testing "exchange registers HL <-> DE"
    (let [ r (assoc sut/init-registers :D 0x01 :E 0x02 :H 0x03 :L 0x04)
          m [0xEB]
          io []
          [r-out m-out io-out] (sut/step r m io) ]
      (t/is (= 0x01 (r-out :H)))
      (t/is (= 0x02 (r-out :L)))
      (t/is (= 0x03 (r-out :D)))
      (t/is (= 0x04 (r-out :E)))
      (t/is (= 0x01 (r-out :PC)))
      ))
         
  )        


src/emulators/intel8080.clj
(ns emulators.intel8080)

;; a library for emulating the Intel 8080 CPU from the 70s.
;; basically takes a state object consiting or registers, memory and
;; and performs the desired action on it.


(def init-registers
  "Returns a map of the CPU registers at startup (or after a reset"
  {
   :A 0 :FL 0
   :B 0 :C 0
   :D 0 :E 0
   :H 0 :L 0
   :SP 0
   :PC 0
   })

(def register-map { 07 :A 00 :B 01 :C 02 :D 03 :E 04 :H 05 :L } )

(def register-pair-map { 00 [:B :C] 01 [:D :E] 02 [:H :L] 03 [:SP] })


;;--------------------------------------------------------------------
;; Helper Functions
;;--------------------------------------------------------------------

(defn- ddd-reg [instr] (register-map (bit-shift-right (bit-and 2r00111000 instr) 3)))
(defn- sss-reg [instr] (register-map (bit-and 2r00000111 instr)))
(defn- get-rp  [instr]  (register-pair-map (bit-shift-right (bit-and 0x30 instr) 4)))

(defn- reg-op [register-set register op value]
  (assoc register-set register (op value (register-set register))))


(defn- to-long [hi-byte low-byte] (+ (* 256 hi-byte) low-byte))
(defn- get-long [register labelH labelL ]
  (to-long (register labelH) (register labelL)))


(defn- instr3 [mem addr]
  [ (mem addr) (mem (+ 1 addr)) (mem (+ 2 addr)) ])


(defn- incr-pc
  "Increments the value of the program counter"
  ( [regs val]
   (reg-op regs :PC + val))

  ( [regs]
   (incr-pc regs 1))

)

;;------------------------------
;; Data Transfer Group
;;------------------------------
(defn- mov-r1-r2 [ regs mem io ]
  (let [ instr (mem (regs :PC))
        ddd   (ddd-reg instr)
        sss   (sss-reg instr)]
    (list (assoc regs ddd (regs sss) :PC (+ 1 (regs :PC)) mem io))))


(defn- mov-r-n [regs mem io]
  (let [ addr (get-long regs :H :L)
         ddd  (ddd-reg (mem (regs :PC)))]
    (list
     (assoc regs ddd (mem addr) :PC (+ 1 (regs :PC)))
     mem
     io)))

(defn- mov-n-r [regs mem io]
  (let [ addr (get-long regs :H :L)
        sss  (sss-reg (mem (regs :PC)))
        ]
    (list
     (incr-pc regs)
     (assoc mem addr (regs sss))
     io)))


(defn- mvi-r-d [regs mem io]
  (let [ ddd (ddd-reg (mem (regs :PC)))
        val (mem (+ 1 (regs :PC)))
        ]
    (list
     (assoc regs ddd val :PC (+ 2 (regs :PC)))
     mem
     io)))
        
(defn- mvi-m-d [regs mem io]
  (let [ addr (get-long regs :H :L)
        val  (mem (+ 1 (regs :PC)))
        ]
    (list
     (incr-pc regs 2)
     (assoc mem addr val)
     io)))


(defn- lxi-rp-d [regs mem io]
  (let [ addr (regs :PC)
        pair (get-rp (mem addr))
        low-byte (mem (+ 1 addr))
        hi-byte  (mem (+ 2 addr))]
    (if (= [:SP] pair)
      (list
       (assoc regs :SP (to-long hi-byte low-byte) :PC (+ 3 addr))
       mem
       io)
      (list
       (assoc regs (first pair) hi-byte (second pair) low-byte :PC (+ 3 addr))
       mem
       io)
      )))


(defn- lda-m [regs mem io]
  (let [ addr (regs :PC)
        mem-addr (to-long (mem (+ 2 addr)) (mem (+ 1 addr))) ]
    (list
     (assoc regs :A (mem mem-addr) :PC (+ 3 addr))
     mem
     io)))

(defn- sta-m [regs mem io]
  (let [ [instr low-byte high-byte] (instr3 mem (regs :PC)) ]
    (list
     (incr-pc regs 3)
     (assoc mem (to-long high-byte low-byte) (regs :A))
     io)))

(defn- lhld [regs mem io]
  (let [ [instr low-byte high-byte ] (instr3 mem (regs :PC))
        addr (to-long high-byte low-byte) ]
    (list
     (assoc regs :L (mem addr) :H (mem (+ 1 addr)) :PC (+ 3 (regs :PC)))
     mem
     io)))

(defn- shld [regs mem io]
  (let [ [instr low-byte high-byte] (instr3 mem (regs :PC))
        mem-addr (to-long high-byte low-byte) ]
    (list
     (incr-pc regs 3)
     (assoc mem mem-addr (regs :L) (+ 1 mem-addr) (regs :H))
     io)
    ))

(defn- ldax [regs mem io]
  (let [ [labelH labelL] (get-rp (mem (regs :PC)))
        mem-addr (get-long regs labelH labelL) ]
    (list
     (assoc regs :A (mem mem-addr) :PC (+ 1 (regs :PC)))
     mem
     io)))

(defn- stax [regs mem io]
  (let [ [labelH labelL] (get-rp (mem (regs :PC)))
        mem-addr (get-long regs labelH labelL) ]
    (list
     (incr-pc regs 1)
     (assoc mem mem-addr (regs :A))
     io)
    ))
                                             
(defn- xchg [regs mem io]
  (list
   (assoc regs :H (regs :D) :D (regs :H) :L (regs :E) :E (regs :L) :PC (+ 1 (regs :PC)))
   mem
   io)
  )

(defn- error-func [regs mem io]
  (throw (IllegalArgumentException. (str "Unrecognized instruction " (mem (regs :PC))))))


(defn instruction-dispatcher
  "Returns the function appropriate for processing instruction."
  [instr]
  (cond
    (and (= 0x40 (bit-and 0xC0 instr)) (not (= 0x30 (bit-and 0x38 instr))) (not (= 0x06 (bit-and 0x07 instr)))) mov-r1-r2
    (and (= 0x46 (bit-and 0xC7 instr)) (not (= 0x30 (bit-and 0x38 instr)))) mov-r-n
    (and (= 0x70 (bit-and 0xF8 instr)) (not (= 0x06 (bit-and 0x07 instr)))) mov-n-r
    (and (= 0x06 (bit-and 0xC7 instr)) (not (= 0x30 (bit-and 0x38 instr)))) mvi-r-d
    (= 0x36 instr) mvi-m-d
    (= 0x01 (bit-and 0xCF instr)) lxi-rp-d
    (= 0x3A instr) lda-m
    (= 0x32 instr) sta-m
    (= 0x2A instr) lhld
    (= 0x22 instr) shld
    (or (= 0x0A instr) (= 0x1A instr)) ldax
    (or (= 0x02 instr) (= 0x12 instr)) stax
    (= 0xEB instr) xchg
    :else error-func))


  
(defn step
  "Executes the current instruction on the CPU returning new register, memory and io states"
  [ regs mem io ]
  (let [instruction (mem (regs :PC))]
    ((instruction-dispatcher instruction) regs mem io)))

;; collection of instructions



